messaging/src/main/java/org/axonframework/commandhandling/gateway/AbstractRetryScheduler.java

```diff
@@ -146,6 +173,106 @@ public boolean scheduleRetry(CommandMessage commandMessage,
+         * existing predicate chain and replaces it with the provided predicate.
+         * <p/>
+         * Provided {@code nonTransientFailurePredicate} checks whether a failure is transient (predicate returns
+         * {@code false}) or not (predicate returns {@code true}). Non-transient failures are never retried.
+         *
+         * @param nonTransientFailurePredicate a {@link Predicate} (accepting a {@link Throwable} parameter) used for
+         *                                     testing transiency of provided failure
+         * @return the current Builder instance, for fluent interfacing
+         */
+        public B nonTransientFailurePredicate(Predicate<Throwable> nonTransientFailurePredicate) {
+            assertNonNull(nonTransientFailurePredicate, "Predicate may not be null");
+            this.nonTransientFailurePredicate = nonTransientFailurePredicate;
+
+            // noinspection unchecked
+            return (B) this;
+        }
+
+        /**
+         * Resets the {@link Predicate} chain used for checking non-transiency of a failure. Throws away any previously
+         * existing predicate chain and replaces it with the provided predicate.
+         * <p/>
+         * Provided {@code nonTransientFailurePredicate} checks whether a failure is transient (predicate returns
+         * {@code false}) or not (predicate returns {@code true}). Non-transient failures are never retried.
+         * <p/>
+         * In essence, this is just a failure typed variant of {@link #nonTransientFailurePredicate(Predicate)} method.
+         *
+         * @param failureType concrete class of the failure to handle
+         * @param nonTransientFailurePredicate a {@link Predicate} (accepting a {@code failureType} parameter) used
+         *                                     for testing transiency of provided failure
+         * @param <E> type of the failure to handle
+         * @return the current Builder instance, for fluent interfacing
+         */
+        public <E extends Throwable> B nonTransientFailurePredicate(Class<E> failureType, Predicate<? super E> nonTransientFailurePredicate) {
+            assertNonNull(failureType, "Class of failure type may not be null");
+            assertNonNull(nonTransientFailurePredicate, "Predicate may not be null");
+
+            //noinspection Convert2MethodRef
+            Predicate<E> typeCheckPredicate = (failureAtRuntime) -> failureType.isInstance(failureAtRuntime);
+
+            // noinspection unchecked
+            this.nonTransientFailurePredicate = ((Predicate<Throwable>)typeCheckPredicate.and(nonTransientFailurePredicate));
+
+            // noinspection unchecked
+            return (B) this;
+        }
+
+        /**
+         * Adds additional predicate into the predicate chain for checking non-transiency of a failure.
+         * <p/>
+         * Provided {@code nonTransientFailurePredicate} is set on the beginning of the predicate chain with
```

Comment:
Curious: why set the new entry as the first entry in the predicate check?

Reply:
It is just easier for me to reason about it in this way. Say "I will first check if my specific predicate will return true, and if it doesn't then will leave the decision to the default" instead "If default fails, then check with my specific predicate". I kinda like to go from specifics to more general even if it doesn't really matter in this case. As said, it is just easier for me to think about it. And it is in Javadoc to give a hint to the user, if needed.

Reply:
Gotcha, sounds reasonable to me. As long as it's documented, we should be good. Thanks for the insights @dmurat!